types/prefs/struct_list.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package prefs
+
+import (
+	"reflect"
+	"slices"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"tailscale.com/types/opt"
+	"tailscale.com/types/ptr"
+	"tailscale.com/types/views"
+)
+
+// StructList is a preference type that holds zero or more potentially mutable struct values.
+type StructList[T views.Cloner[T]] struct {
+	preference[[]T]
+}
+
+// StructListOf returns a [StructList] configured with the specified value and [Options].
+func StructListOf[T views.Cloner[T]](v []T, opts ...Options) StructList[T] {
+	return StructList[T]{preferenceOf(opt.ValueOf(deepCloneSlice(v)), opts...)}
+}
+
+// StructListWithOpts returns an unconfigured [StructList] with the specified [Options].
+func StructListWithOpts[T views.Cloner[T]](opts ...Options) StructList[T] {
+	return StructList[T]{preferenceOf(opt.Value[[]T]{}, opts...)}
+}
+
+// Clone returns a copy of l that aliases no memory with l.
+func (l *StructList[T]) Clone() *StructList[T] {
+	res := ptr.To(*l)
+	if v, ok := l.s.Value.GetOk(); ok {
+		res.s.Value.Set(deepCloneSlice(v))
+	}
+	return res
+}
+
+// Equal reports whether l and l2 are equal.
+// If the template type T implements an Equal(T) bool method, it will be used
+// instead of the == operator for value comparison.
+// If T is not comparable, it reports false.
+func (l StructList[T]) Equal(l2 StructList[T]) bool {
+	if l.s.Metadata != l2.s.Metadata {
+		return false
+	}
+	v1, ok1 := l.s.Value.GetOk()
+	v2, ok2 := l2.s.Value.GetOk()
+	if ok1 != ok2 {
+		return false
+	}
+	if ok1 != ok2 {
+		return false
+	}
+	return !ok1 || slices.EqualFunc(v1, v2, comparerFor[T]())
+}
+
+func deepCloneSlice[T views.Cloner[T]](s []T) []T {
+	c := make([]T, len(s))
+	for i := range s {
+		c[i] = s[i].Clone()
+	}
+	return c
+}
+
+type equatable[T any] interface {
+	Equal(other T) bool
+}
+
+func comparerFor[T any]() func(a, b T) bool {
+	switch t := reflect.TypeFor[T](); {
+	case t.Implements(reflect.TypeFor[equatable[T]]()):
+		return func(a, b T) bool { return any(a).(equatable[T]).Equal(b) }
+	case t.Comparable():
+		return func(a, b T) bool { return any(a) == any(b) }
+	default:
+		return func(a, b T) bool { return false }
```

Comment:
should this be a panic? otherwise `Equal` would return incorrect results

Reply:
I'm unsure and don't have a strong opinion on this one.

I generally prefer a panic in situations like this, but I've been told on a few occasions that we prefer not to panic unless absolutely necessary. There're also cases when somewhat similar functions return `false` for incomparable values (e.g. `reflect.DeepEqual` when comparing non-nil functions). So in this case I chose to document it rather than to panic: https://github.com/tailscale/tailscale/blob/80bf97be3ac073642f4a52b53d0e700c131da52e/types/prefs/struct_list.go#L41-L45

Happy to change to a panic if you think it's better!

Reply:
Updated it to panic.